### PR TITLE
Feature/tuple syntax

### PR DESCRIPTION
**Pull request by alexandra-hanson**
I have added syntax for tuples as an extension, and done so inductively. We forward a tuple `(x, y, z)` to `pair(x, pair(y, z))`, as described in #399.

I would like feedback on if I have provided sufficient test cases, and if tuple behavior is as desired.

### CODE REVIEW

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (krame505) [inline grammars/silver/compiler/extension/tuple/Tuple.sv:19]: Minor style note: we typically indent by 2 spaces throughout our Silver code base.

**Review by krame505 — COMMENTED** [review]
What you have so far looks great - I will approve this once you have the type expression syntax finished.  

Also for completeness you could/probably should add `()` as a type that forwards to `Unit` and as a pattern/expression that forwards to `unit()`.

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (krame505) [inline grammars/silver/compiler/extension/tuple/Tuple.sv:40]: I would maybe just call this `tupleExpr`.

**Comment by krame505** [comment]
This is pretty good as-is, however currently tuple types would still show up in error messages as the underlying chain of `Pair` types.  This could be fixed by adding a `tupleType :: (Type ::= [Type])` production that forwards to the expected chain of `Pair` types.  `tupleType` would override its `unparse` to use the parenthesized representation, while `tupleTypeExpr`  and `tupleExpr` would override their types to be `tupleType` applied to the appropriate arguments.

**Comment by ericvanwyk** [comment]
Yes, nicer type error messages would be nice to have.

**Review by krame505 — APPROVED** [review]

**Comment by krame505** [comment]
You might also consider trying to use this in the Silver compiler, e.g. https://github.com/melt-umn/silver/blob/develop/grammars/silver/compiler/metatranslation/Translation.sv#L83 and a few other places in the that file involve pathological chains of `Pair` types.

File: grammars/silver/compiler/metatranslation/Translation.sv
Comment (alexandra-hanson) [inline grammars/silver/compiler/metatranslation/Translation.sv:94]: The second element of the tuple is still accessed by trans.snd.fst. Is that confusing/is there a better way for this access to occur?

Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/metatranslation/Translation.sv:94]: As I recall (and it looks like it says this at the bottom of the tuple issue as well), we agreed it would be a good idea to use `.<int>` to access elements of a tuple.  For your example here, that would be `trans.2`.

Comment (alexandra-hanson) [reply grammars/silver/compiler/metatranslation/Translation.sv:94]: My bad! I will work on that now.

**Comment by krame505** [comment]
@alexandra-hanson is this ready to be merged?  Or did you want to finish up the new selector syntax in this branch first, before merging this?

**Comment by ericvanwyk** [comment]
I think it might be good to wait.  @alexandra-hanson and I chatted today and I had some questions about how things were done - mostly down to my ignorance on things.  We may corner you, @krame505, for a conversation next week sometime.

**Comment by alexandra-hanson** [comment]
I have added a `/tuple/README.md` to attempt to explain how tuples work. Probably the documentation is a bit excessive, but I needed to write it down to explain how it worked to myself. There also could be things I am missing. I have added notes that are questions and thoughts.

Probably the most pertinent section is about types. This contains the most thoughts/questions also.

File: grammars/silver/compiler/extension/tuple/README.md
Comment (RandomActsOfGrammar) [inline grammars/silver/compiler/extension/tuple/README.md:33]: The default production is shorthand for writing the same equation on every production other than the ones we really care about.  If we have `top.attr = 0;` in the default production, this is the equation in all the productions in the grammar where we *don't* write an equation.

By putting `top.tupleElems = [top];` in the default production, you're saying that the list of types in the "tuple" represented by that type is simply the type itself.

File: grammars/silver/compiler/extension/tuple/README.md
Comment (RandomActsOfGrammar) [inline grammars/silver/compiler/extension/tuple/README.md:32]: There will be a value of `tupleElems` for each tree of type `Type`.  If you look at the equation you have on `appType` for `tupleElems`, you will see that you only do something interesting in the case that it is a pair; otherwise, you do the same as every other type, where you have `top.tupleElems = [top];`.

Comment (krame505) [reply grammars/silver/compiler/extension/tuple/README.md:32]: Not sure if this is what you are concerned about, but remember that attributes are computed lazily - so this attribute on `Type` won't add performance overhead or anything where it isn't explicitly demanded by the tuple extension.

File: grammars/silver/compiler/extension/tuple/README.md
Comment (RandomActsOfGrammar) [inline grammars/silver/compiler/extension/tuple/README.md:39]: This lets substitutions be done without destroying the original structure.  If `substituted` were done by forwarding, the value of `substituted` would have the structure of the pair types to which this forwards rather than the tuple type it originally was.

**Comment by RandomActsOfGrammar** [comment]
I like the idea of having a `README.md` for each extension to provide a guide for future development on it.

Your current version is good while you're still figuring things out, but a lot of it is just translating the code into English.  The final version can be shorter and include less detail.  For example, once this is merged, people can see that `Silver_Expr` is used to create the forwarded-to pairs; it isn't an unusual enough feature to be pointed out in documentation.

File: grammars/silver/compiler/extension/tuple/README.md
Comment (alexandra-hanson) [reply grammars/silver/compiler/extension/tuple/README.md:32]: Ah okay!! Yes, that makes a lot of sense. Thank you both.

File: grammars/silver/compiler/metatranslation/Translation.sv
Comment (alexandra-hanson) [reply grammars/silver/compiler/metatranslation/Translation.sv]: Yeah, with what I haven't pushed yet -- trying to add the selector syntax -- it would show up with errors on translation during compilation and not notify me anything about what was wrong with my tuple syntax. I also had to merge in the changes from develop and there were conflicts so I thought it would just be easier to revert it. I will change it back when it is working. Sorry if this is not ideal.

**Comment by ericvanwyk** [comment]
I agree - these README.md files are helpful.  Especially in an initial form that tracks the thoughts and questions that arise.  When Allie and I last spoke we thought something like this would be good to sort things out and get some feedback.  Thanks all.

**Comment by alexandra-hanson** [comment]
@krame505 and @RandomActsOfGrammar Thank you both! Your comments were really helpful. I think that has cleared up where I was confused with the way types are implemented. It was helpful for me to write up the code into English to get my thoughts in order and figure out where my questions were but I will certainly shorten the README.md once I'm done.

File: grammars/silver/compiler/extension/implicit_monads/Case.sv
Comment (alexandra-hanson) [inline grammars/silver/compiler/extension/implicit_monads/Case.sv:163]: I wanted to change these `.fst` and `.snd` accesses to tuple accesses, such that we say `subcall.1` and `subcall.2`; however, when I attempted to do so, I received the error:

`Case.sv:161:19: error: Argument 2 of function 'silver:core:cons' expected [(silver:compiler:definition:type:Type, silver:compiler:definition:core:Expr, String)] but argument is of type Decorated ([(silver:compiler:definition:type:Type, silver:compiler:definition:core:Expr, String)], silver:core:List<silver:compiler:definition:core:Expr>)
Case.sv:162:66: error: Invalid tuple selector index.`

I added a test just to make sure that the tuple selector syntax worked on 2 element tuples. I don't understand exactly what this error means/how I should approach fixing this.

File: grammars/silver/compiler/extension/implicit_monads/Case.sv
Comment (krame505) [inline grammars/silver/compiler/extension/implicit_monads/Case.sv:138]: This could just be `(Type, Expr, String)`

File: grammars/silver/compiler/extension/implicit_monads/Case.sv
Comment (krame505) [reply grammars/silver/compiler/extension/implicit_monads/Case.sv:163]: Looks like on line 163 you have `subcall.snd` instead of `e :: subcall.snd`?

Also you probably need to add an equation for `tupleElems` on `listType`/listCtrType`, since the error message is showing `silver:core:List<...>` instead of `[...]`

File: grammars/silver/compiler/extension/tuple/README.md
Comment (alexandra-hanson) [inline grammars/silver/compiler/extension/tuple/README.md:5]: Not sure if this information should go in the README or on the website. I want to give a brief overview of how tuples work but I'm not sure if this is brief enough here. In this section I could just point to the website to explain usage?

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (alexandra-hanson) [inline grammars/silver/compiler/extension/tuple/Tuple.sv:33]: Is this correct & do I need more information to explain this than this comment?

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (alexandra-hanson) [inline grammars/silver/compiler/extension/tuple/Tuple.sv:53]: Is it correct to say that we are "undecorating" t. I'm not sure if this is the correct terminology.

File: grammars/silver/compiler/extension/tuple/Type.sv
Comment (alexandra-hanson) [inline grammars/silver/compiler/extension/tuple/Type.sv:56]: Why do we do this with `top.substituted` and `top.flatRenamed` again?

File: grammars/silver/compiler/extension/tuple/README.md
Comment (krame505) [reply grammars/silver/compiler/extension/tuple/README.md:5]: I'm not sure how I feel about having README.md files scattered in the code base - this would be the only non-`.sv` file under `grammars/`, as far as I know.  The info before the "Implementation" section here (don't to mention patterns too!) could fit on the website under http://melt.cs.umn.edu/silver/ref/ext/.

The rest of the implementation description can probably move into the source code as doc comments in appropriate places - so this would (eventually) end up on the website as well if we decide to put up generated docs for the compiler internals.  

The way that this is implemented doesn't really have `,` as an infix operator, by the way - if it were so you could write just `[String, Integer]` without the parentheses.  But it can be helpful to think about that way, I guess.

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (krame505) [reply grammars/silver/compiler/extension/tuple/Tuple.sv:33]: Umm, that's not really super clear.  The key thing is that computing the specialized `tupleType` from `forward.typerep` is cleaner than performing type checking on the `TupleList`.  `performSubstitution` is needed because `forward.typerep` isn't instantiated into a chain of pairs until `upSubst` is applied.

File: grammars/silver/compiler/extension/tuple/Tuple.sv
Comment (krame505) [reply grammars/silver/compiler/extension/tuple/Tuple.sv:53]: Not exactly, typically by that we mean turning a decorated tree into an undecorated term.  Here we just need to get at the underlying `tupleType` from a type that might be inferred to be decorated by default.

File: grammars/silver/compiler/extension/tuple/Type.sv
Comment (krame505) [reply grammars/silver/compiler/extension/tuple/Type.sv:56]: We want to avoid transforming away the `tupleType` and turning it back into a chain of `Pair`s when performing substitutions, like we discussed today with lists.

File: grammars/silver/compiler/extension/tuple/README.md
Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/extension/tuple/README.md:5]: I think the benefit of having `README.md` files in extension directories is to provide documentation on that extension for Silver developers, so if anyone needs to make changes to the extension they have a quick introduction to it.

With that interpretation, the part before `Implementation` doesn't really belong here, since that is more for users of the extension than for developers of it.

Comment (krame505) [reply grammars/silver/compiler/extension/tuple/README.md:5]: > I think the benefit of having README.md files in extension directories is to provide documentation on that extension for Silver developers, so if anyone needs to make changes to the extension they have a quick introduction to it.

Yes, but that is what doc comments are for.  There are a number of comments in the Silver compiler source that should become generated doc comments once that is working.  My main issue with using `README.md` files is that these won't be included in the generated documentation.

Comment (alexandra-hanson) [reply grammars/silver/compiler/extension/tuple/README.md:5]: I can translate the `README.md` into doc comments but I'm not entirely sure if there is a consensus on this so I'm not sure how I should proceed.

Comment (krame505) [reply grammars/silver/compiler/extension/tuple/README.md]: If @ericvanwyk thinks that we should have README files in extension grammars then I guess this is fine.  If these were doc comments instead they would still be in the source tree and also get added to the website.  But that can always be fixed later if/when we start putting generated docs from the `silver:compiler` grammars on our website.

Comment (ericvanwyk) [reply grammars/silver/compiler/extension/tuple/README.md]: Every repo should have a README since this might be some reader's first entrance into that material.  It might be that the best thing to have in the README are not much more than pointers to the Silver web page for how to use, in this case, the extension, and to the generated doc-comments that describe the implementation of the specifications in the repo.  In the case of this one I think it makes sense to say what the purpose of the extension is, a little bit about what it provides and then pointers to the relevant web pages that have the documentation. We do want to avoid multiple copies of large bits of documentation - but a paragraph or two that lives here in the README and also somewhere doesn't seem unreasonable. At least until we have the Silver documentation sorted out. What is here now might be more than we need - but again - we need to have the other documentation sorted out before we judge what is here as being too much or too little.

**Pull request merged**
